lib/blocks.php

```diff
@@ -296,3 +296,219 @@ function gutenberg_enqueue_editor_block_styles_assets() {
+		$has_custom_text_color = array_key_exists( 'style', $attributes )
+		&& array_key_exists( 'color', $attributes['style'] )
+		&& array_key_exists( 'text', $attributes['style']['color'] );
+
+		// Apply required generic class.
+		if ( $has_custom_text_color || $has_named_text_color ) {
+			$color_settings['css_classes'][] = 'has-text-color';
+		}
+		// Apply color class or inline style.
+		if ( $has_named_text_color ) {
+			$color_settings['css_classes'][] = sprintf( 'has-%s-color', $attributes['textColor'] );
+		} elseif ( $has_custom_text_color ) {
+			$color_settings['inline_styles'] .= sprintf( 'color: %s;', $attributes['style']['color']['text'] );
+		}
+	}
+
+	// Link Colors.
+	if ( in_array( 'link-color', $supports, true ) ) {
+		$has_link_color = array_key_exists( 'style', $attributes )
+		&& array_key_exists( 'color', $attributes['style'] )
+		&& array_key_exists( 'link', $attributes['style']['color'] );
+		// Apply required class and style.
+		if ( $has_link_color ) {
+			$color_settings['css_classes'][] = 'has-link-color';
+			// If link is a named color.
+			if ( strpos( $attributes['style']['color']['link'], 'var:preset|color|' ) !== false ) {
+				// Get the name from the string and add proper styles.
+				$index_to_splice                  = strrpos( $attributes['style']['color']['link'], '|' ) + 1;
+				$link_color_name                  = substr( $attributes['style']['color']['link'], $index_to_splice );
+				$color_settings['inline_styles'] .= sprintf( '--wp--style--color--link:var(--wp--preset--color--%s);', $link_color_name );
+			} else {
+				$color_settings['inline_styles'] .= sprintf( '--wp--style--color--link: %s;', $attributes['style']['color']['link'] );
+			}
+		}
+	}
+
+	// Background Colors.
+	if ( in_array( 'background-color', $supports, true ) ) {
+		$has_named_background_color  = array_key_exists( 'backgroundColor', $attributes );
+		$has_custom_background_color = array_key_exists( 'style', $attributes )
+		&& array_key_exists( 'color', $attributes['style'] )
+		&& array_key_exists( 'background', $attributes['style']['color'] );
+
+		// Apply required background class.
+		if ( $has_custom_background_color || $has_named_background_color ) {
+			$color_settings['css_classes'][] = 'has-background';
+		}
+		// Apply background color classes or styles.
+		if ( $has_named_background_color ) {
+			$color_settings['css_classes'][] = sprintf( 'has-%s-background-color', $attributes['backgroundColor'] );
+		} elseif ( $has_custom_background_color ) {
+			$color_settings['inline_styles'] .= sprintf( 'background-color: %s;', $attributes['style']['color']['background'] );
+		}
+	}
+
+	// Gradients.
+	if ( in_array( 'background', $supports, true ) ) {
+		$has_named_gradient  = array_key_exists( 'gradient', $attributes );
+		$has_custom_gradient = array_key_exists( 'style', $attributes )
+		&& array_key_exists( 'color', $attributes['style'] )
+		&& array_key_exists( 'gradient', $attributes['style']['color'] );
```

Comment:
Many of these checks across gradient/background/link/text seem to be very similar. Maybe it would be useful to create a helper function?

(I see that typography also looks extremely similar to these checks)

Reply:
apparently we can just use `isset( $attributes['style']['color']['text'] );` as it doesn't error if the earlier properties aren't set.  That was a neat tidbit of info to learn - https://github.com/WordPress/gutenberg/pull/23007#discussion_r441261286